Angular upgrade milestone 1

fuzzybair opened the pull request:
I have completed basic login, navigation and the dashboard page. All other pages are currently just placeholders for future work.  After login you can switch between the existing interface and the new interface through the menu on ether side.

Screen shots:
Light
![image](https://user-images.githubusercontent.com/5714774/126083044-cf2daf9b-3e8f-4aa3-9279-22594408f08e.png)
![image](https://user-images.githubusercontent.com/5714774/126083076-4e516f42-7231-4cb9-bba6-733557137c75.png)
![image](https://user-images.githubusercontent.com/5714774/126083083-08ede5b4-ee99-46b6-96ae-98db7e75ca70.png)

Dark
![image](https://user-images.githubusercontent.com/5714774/126082948-7af585cb-770c-4a9c-a4c2-a8d77583a5e5.png)
![image](https://user-images.githubusercontent.com/5714774/126082818-1c4ab865-3158-456c-a51f-13b5a5fe5b65.png)
![image](https://user-images.githubusercontent.com/5714774/126082834-43d456d3-d4a4-44df-9d14-a9f0eeaf0ca9.png)

hexparrot commented:
Are there `/ui` files absent from this branch?  I haven't been able to get this to work, as switching to the new UI via the interface returns me `Cannot GET /ui`. There does not appear to be a directory of that name, was it accidentally left out?

fuzzybair commented:
The files in /ui are compiled artifacts from Angular, the [Dockerfile](https://github.com/hexparrot/mineos-node/pull/428/commits/49c41e2b0ccaae446e9e09d6c2bc2fdfdb31aacd#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R43) and [update_webui.sh](https://github.com/hexparrot/mineos-node/pull/428/commits/49c41e2b0ccaae446e9e09d6c2bc2fdfdb31aacd#diff-da164c53ab3a02275e9c1b31352643f427556caa32f3bd5ea8bbe0c894d73feeR40) scripts will do that for you but as they are doing a git checkout from master rather than using the local copy they wont work until after everything is merged, or unless you modify the files.  As for testing prior to merge just run
```
cd [mineos git path]/mineos-app
npm install;
npm run build;
```
looks like there is a minor issue in the Dockerfile it is just calling `ng build` instead of `npm run build` as a result it wont work. I will get a fix up for that.

fuzzybair commented:
I know life gets busy and this is a large pull request. I am just wondering what is the status? Do you need anything updated?

hexparrot commented:
Sorry for the delay; I'll look at this today

fuzzybair commented:
The attached screen shots are not mockups those are actually functional. But to your point most of the pages are not, they all look something like 
![image](https://user-images.githubusercontent.com/5714774/131275195-9a650489-f7e6-4632-b29c-ebe7faef9dfa.png)
I am ok if we want to work off a branch. However I think future pages will likely need to make modifications to the existing backend, and I am worried that will become an issue if it gets stale. 

I did notice that socket.io was not interoperable between 1.x and 4.x however I could not find anything that said it was not API compatible. I updated the package and clicked through all of the links in the existing UI that I could find and everything seemed to work as expected. I do no believe there are any required changes needed for the socket.io upgrade, there may be new API's that you would like to take advantage of but I did not look into that.

I highly recommend [Tour of Heroes app and tutorial](https://angular.io/tutorial) from the Angular team to anyone that is new to Angular. You may find [AngularJS to Angular concepts](https://angular.io/guide/ajs-quick-reference) useful. I have used it a lot for translating what I know back to AngularJS, but the intended audience is existing AnularJS developers starting into the new Angular.

As for the layout of the code I am a firm believer in Test driven development the `*.spec.ts` files are unit tests written using Jasmine, which is the default test framework for Angular. see [Angular Testing](https://angular.io/guide/testing) for more information. The other `*.ts` files are [TypeScript](https://www.typescriptlang.org/) files (the default language for Angular), this is the actual application code. TypeScript is a superset of JavaScript so you can use JavaScript methods you are used to. The angular cli transpiles the typescript to JavaScript for the browser. As a result all of the files in `/ui/` will be `*.js`. TypeScript has many development time benefits, through IDE integration, error checking, etc. The model files are just TypeScript representations of JSON objects, nothing special.

As for the "basic mechanism for transferring information from the server to the client" as a general rule that will be JSON typically done over REST calls (as you will see with most examples online). MineOS as I understand it uses socket.io and websockets for most, if not all of its communication to the backend. I see no reason this approach can't continue to work. If we decide to deviate from that I think it would be best to do it after the existing UI is retired.

To help get things running here are some steps I just worked out using the docker image as a base (to avoid any pre-assumptions in my development environment). 

_(not sure why build-essential is needed but without it the backend does not seem to work for me)_
````
docker run --rm -it -e USE_HTTPS="false" -e USER_PASSWORD="test" -e SERVER_PORT=80 -p 80:80 hexparrot/mineos /bin/bash
apt-get update && apt-get install -y build-essential
git config --global user.email "you@example.com"
git config --global user.name "Your Name"
cd /usr/games/minecraft
````


I would assume if you have a working development environment the above is probably not necessary the key things you need are as follows (this assumes you have master checked out):

1. Merge this PR into an up to date master
````
git pull
git fetch origin pull/428/head:pr-428
git merge pr-428
````
2. Build the new UI code. _(the new UI uses a separate package.json and node_modules folder)_ 
````
cd mineos-app/
npm install --no-spin --unsafe-perm
npm run build
cd ..
````
3. Start the MineOS backend with existing UI _(this should be the same as you have done in the past)_
````
rm -fr node_modules
npm install --no-spin --unsafe-perm
npm run start
```` 

As for you working on this effort, while I am not opposed in any way, I am also not expecting it. I still feel this is not going to be something I can't manage, and I am happy to do it. As the project owner I was really just looking to you for approval and making sure I am not introducing regressions in the exiting UI. 

A future feature enhancement may be to have some Selenium/cypress tests to validate UI functionality. This would off load the strain on you in testing, or the risk of introducing regressions. And also help with onboarding new developers that may not have used all of the features like myself.

fuzzybair commented:
I have looked through all the "Breaking Changes" for socket.io from 1.7 through the latest 4.x here is the summary. https://github.com/socketio/socket.io/releases

Breaking Change | Fix needed
-------------------|--------------
utf-8 encoding in engine.io-parser | Nothing – given The MineOS UI is the only client it will upgrade with the server so this is only an issue if there are third party/additional clients for the MineOS server
the socket id on the client match the id on the server-side | Nothing – no references in MineOS to the internal client socket id only session id 
the Socket#use() method is removed | [only one reference](https://github.com/hexparrot/mineos-node/blob/master/webui.js#L106) This may need a fix – This may explain some inconsistent login issues I have been having that I thought was related to my switching between UI's.
Socket#join() and Socket#leave() | Nothing – no references found in MineOS
io.to(...) now returns an immutable operator | Unsure here – given the previous behavior described io.to(“server1”).emit(msg1) then calling io.to(“server2”).emit(msg2) msg2 gets sent to both server1 and server2. I don’t think that was ever your intention based on how I am reading the code. If I am wrong there may be some significant changes here.

I will lookin to the one call to use() and see if it needs changed. out side of that I see nothing that would stop MineOS from jumping to the latest 4.x version.

fuzzybair commented:
I was just re-reading the documentation on socket.io and the git commit related to the change listed above. `io.use()` was not the method in question it was `socket.use()` which is not used in MineOS and for further compatibility version 3.0.0 removed it and 3.0.5 added it back using the 1.6.x version 
![image](https://user-images.githubusercontent.com/5714774/133002874-f8ea7a96-36f5-47d8-9c93-200b294755f2.png)
to that end there is no breaking change in socket.io that impacts the existing MineOS UI or server. However the newer Angular packages are built linked to the 4.x so it would be difficult to attempt to downgrade to 1.x for what appears to me to have no added benefit or impact to the code.

fuzzybair commented:
I know you probably have not had a chance to respond to my previous comments but I just wanted to see if my above comments resolved your concerns or if there is more you need from me. From your comment these are the concerns I see:


> “Can you let me know, then, how functional is this UI for you?” 

Answered in more detail above but to summarize the existing UI is still fully functional the new UI is just showcasing that I can receive and display data from the MineOS NodeJS server.

> “The Socket.io update invalidates all my existing code, which breaks from the major versions going 1 -> 4.” 

More details above but the short version is.  The Socket.io update does not invalidate any existing code. Socket.io (as with many other JavaScript libraries) follows “Semantic Versioning” so a major version change does not indicate a new API just that it had some (generally small) change that is not 100% backwards compatible see semver.org

>  “I hope that we can get a basic variable from the server to update on the client.”. 

Due to how MineOS is written this involves a few steps First authentication is required, all endpoints and webscoket messages that return data require authentication. Second all meaningful data is sent using websockets so a websocket implementation is required, Third a page to display the data. This is effectively what the dashboard page is. This PR simply represents the ability to show meaningful data from the back-end to a new UI the only thing added in addition to what is minimally required to show what you are requesting is basic navigation for all of the up coming pages. Providing the place holder pages allows for more than one developer to work without causing a lot of conflicts, I have a friend that is looking to help as well. Pages can be assigned out as tickets so we can work together without conflicts and/or regular synchronization of our efforts.

> “Making correcting the onslaught of new upgrades that will be required when Angular obsoletes my angular.js legacy code” 


I am planning on supporting the UI efforts beyond the initial implementation, This should give you plenty of time to come up to speed with Angular. I would also be happy to put together some developer documentation in the form of Readme’s and UML diagrams if you are interested.

If you would still like _"a dynamic 'hello world'”_ I can put in a separate PR for that it should only take me a couple hours. But it will not be anything that would be useful in the long run of the upgrade. The design of this page would be as follows, A new unauthenticated i.e. anonymous endpoint added that would do something like return server time. A simple REST call and HTML page that shows `<h1>The server time is {{ time }}</h1>`. This will show simple communication with NodeJS but not in the way that that would be useful for MineOS, for the reasons laid out in the above section related to authentication and websockets.

Have I addressed your concerns, or is there something else I can do to help?
